Har file is not being written to disk when using har dump addon with Mitmdump 4.0.4

artifacts:
1. mitmdump 4.0.2 - downloaded release binaries from the site marked as 4.0.4 (which is a bit odd)
2. har_dump example addon from github repository from '4.x' branch.

##### Steps to reproduce the problem:
1. Try to run 'mitmdump' from cmd, with the mitmdump example addon.
2. No har file is created on exit 

**also reproduced on MacOS.


##### Any other comments? What have you tried so far?
Works perfectly with mitmdump 3.0.4

##### System information
Mitmproxy: 4.0.2 binary
Python:    3.6.5
OpenSSL:   OpenSSL 1.1.0h  27 Mar 2018
Platform:  Windows-10-10.0.17134

Tests for `har_dump.py` are also failing. I fixed simple path problem in #4424 and there are stil errors.

Understanding how to run test in `examples/` is not trivial. Basically `pytest` needs `conftest.py` file with fixtures to be somewhere on the path. I could not find a way to specify location of this file for tests in `examples/`, so I symlinked it.
```
ln -sr test/conftest.py examples
pytest examples/contrib/test_har_dump.py
```

After that, the primary error is still this.
```python
_____________________________ TestHARDump.test_simple _____________________________

self = <test_har_dump.TestHARDump object at 0x7fb35cdb2790>
tmpdir = local('/tmp/pytest-of-anatoli/pytest-10/test_simple0')
tdata = <mitmproxy.utils.data.Data object at 0x7fb35cdbcc70>

    def test_simple(self, tmpdir, tdata):
        with taddons.context() as tctx:
            a = tctx.script(tdata.path("../examples/contrib/har_dump.py"))
            path = str(tmpdir.join("somefile"))
>           tctx.configure(a, hardump=path)

examples/contrib/test_har_dump.py:26: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
mitmproxy/test/taddons.py:107: in configure
    self.options.update(**kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = mitmproxy.options.Options({
    'add_upstream_certs_to_client_chain': False [<class 'bool'>],
    'allow_hosts': [] [t...osts': [] [typing.Sequence[str]],
    'upstream_cert': True [<class 'bool'>],
    'websocket': True [<class 'bool'>]
})
kwargs = {'hardump': '/tmp/pytest-of-anatoli/pytest-10/test_simple0/somefile'}
u = {'hardump': '/tmp/pytest-of-anatoli/pytest-10/test_simple0/somefile'}

    def update(self, **kwargs):
        u = self.update_known(**kwargs)
        if u:
>           raise KeyError("Unknown options: %s" % ", ".join(u.keys()))
E           KeyError: 'Unknown options: hardump'
```

@mhils `mitmproxy.ctx.log` doesn't exist at the time the script is loaded.
```python
        from mitmproxy import ctx
        ns = script.load_script(
            tdata.path(
                "mitmproxy/data/addonscripts/addon.py"
            )
        )
>       ctx.log('yyyyyy')
E       AttributeError: module 'mitmproxy.ctx' has no attribute 'log'
```

--- a/examples/contrib/har_dump.py
+++ b/examples/contrib/har_dump.py
@@ -20,7 +20,7 @@ from datetime import timezone
 
 import mitmproxy
 
-from mitmproxy import connections  # noqa
+from mitmproxy import connection
 from mitmproxy import version
 from mitmproxy import ctx
 from mitmproxy.utils import strutils
@@ -30,7 +30,7 @@ HAR: typing.Dict = {}
 
 # A list of server seen till now is maintained so we can avoid
 # using 'connect' time for entries that use an existing connection.
-SERVERS_SEEN: typing.Set[connections.ServerConnection] = set()
+SERVERS_SEEN: typing.Set[connection.Server] = set()
 
 
 def load(l):

--- a/examples/contrib/har_dump.py
+++ b/examples/contrib/har_dump.py
@@ -53,7 +53,7 @@ def configure(updated):
     })
 
 
-def response(flow):
+def response(flow: mitmproxy.http.HTTPFlow):
     """
        Called when a server response has been received.
     """
@@ -153,7 +153,7 @@ def response(flow):
             "params": params
         }
 
-    if flow.server_conn.connected():
+    if flow.server_conn.connected:
         entry["serverIPAddress"] = str(flow.server_conn.ip_address[0])
 
     HAR["log"]["entries"].append(entry)

--- a/examples/contrib/test_har_dump.py
+++ b/examples/contrib/test_har_dump.py
@@ -20,8 +20,14 @@ class TestHARDump:
         )
 
     def test_simple(self, tmpdir, tdata):
+        # context is needed to provide ctx.log function that
+        # is invoked if there are exceptions
         with taddons.context() as tctx:
             a = tctx.script(tdata.path("../examples/contrib/har_dump.py"))
+            # check script is read without errors
+            assert tctx.master.logs == []
+            assert a.name_value   # last function in har_dump.py
+
             path = str(tmpdir.join("somefile"))
             tctx.configure(a, hardump=path)
             tctx.invoke(a, "response", self.flow())

--- a/examples/contrib/test_har_dump.py
+++ b/examples/contrib/test_har_dump.py
@@ -30,8 +30,8 @@ class TestHARDump:
 
             path = str(tmpdir.join("somefile"))
             tctx.configure(a, hardump=path)
-            tctx.invoke(a, "response", self.flow())
-            tctx.invoke(a, "done")
+            a.response(self.flow())
+            a.done()
             with open(path) as inp:
                 har = json.load(inp)
             assert len(har["log"]["entries"]) == 1
@@ -42,10 +42,8 @@ class TestHARDump:
             path = str(tmpdir.join("somefile"))
             tctx.configure(a, hardump=path)
 
-            tctx.invoke(
-                a, "response", self.flow(resp_content=b"foo" + b"\xFF" * 10)
-            )
-            tctx.invoke(a, "done")
+            a.response(self.flow(resp_content=b"foo" + b"\xFF" * 10))
+            a.done()
             with open(path) as inp:
                 har = json.load(inp)
             assert har["log"]["entries"][0]["response"]["content"]["encoding"] == "base64"
@@ -82,8 +80,8 @@ class TestHARDump:
             f.response.headers["random-junk"] = bytes(range(256))
             f.response.content = bytes(range(256))
 
-            tctx.invoke(a, "response", f)
-            tctx.invoke(a, "done")
+            a.response(f)
+            a.done()
 
             with open(path) as inp:
                 har = json.load(inp)

